File: sdk/python/foodblock/merge.py
"""
FoodBlock Merge — conflict resolution for forked update chains.

When two parties independently update the same block, the chain forks.
This module detects forks, finds common ancestors, and creates merge blocks
to reconcile the divergent states.

Usage:
    conflict = detect_conflict(hash_a, hash_b, resolve)
    merged = merge(hash_a, hash_b, resolve, strategy='a_wins')
"""

from .block import create


def detect_conflict(hash_a, hash_b, resolve):
    """
    Walk both update chains to find a common ancestor and determine if
    the two hashes represent a conflict (fork).

    Args:
        hash_a: hash of one chain head
        hash_b: hash of the other chain head
        resolve: callable (hash) -> block dict or None

    Returns:
        {
            'is_conflict': bool,
            'common_ancestor': hash_str or None,
            'chain_a': [block, ...] from hash_a back to ancestor,
            'chain_b': [block, ...] from hash_b back to ancestor,
        }
    """
    if hash_a == hash_b:
        block = resolve(hash_a)
        return {
            'is_conflict': False,
            'common_ancestor': hash_a,
            'chain_a': [block] if block else [],
            'chain_b': [block] if block else [],
        }

    chain_a = _walk_chain(hash_a, resolve)
    chain_b = _walk_chain(hash_b, resolve)

    hashes_a = {b['hash'] for b in chain_a}
    hashes_b = {b['hash'] for b in chain_b}

    # Find common ancestor: first hash in chain_a that also appears in chain_b
    common_ancestor = None
    for block in chain_a:
        if block['hash'] in hashes_b:
            common_ancestor = block['hash']
            break

    # If no common ancestor found searching from a, try from b
    if common_ancestor is None:
        for block in chain_b:
            if block['hash'] in hashes_a:
                common_ancestor = block['hash']
                break

    # Trim chains to stop at common ancestor
    trimmed_a = _trim_to_ancestor(chain_a, common_ancestor)
    trimmed_b = _trim_to_ancestor(chain_b, common_ancestor)

    is_conflict = (common_ancestor is not None and
                   hash_a not in hashes_b and
                   hash_b not in hashes_a)

    return {
        'is_conflict': is_conflict,
        'common_ancestor': common_ancestor,
        'chain_a': trimmed_a,
        'chain_b': trimmed_b,
    }


def merge(hash_a, hash_b, resolve, state=None, strategy='manual'):
    """
    Create an observe.merge block to reconcile two forked chains.

    Args:
        hash_a: hash of chain A head
        hash_b: hash of chain B head
        resolve: callable (hash) -> block dict or None
        state: optional explicit merged state dict
        strategy: 'a_wins', 'b_wins', or 'manual'

    Returns:
        An observe.merge FoodBlock dict
    """
    block_a = resolve(hash_a)
    block_b = resolve(hash_b)

    if state is None:
        if strategy == 'a_wins':
            state = dict(block_a.get('state', {})) if block_a else {}
        elif strategy == 'b_wins':
            state = dict(block_b.get('state', {})) if block_b else {}
        elif strategy == 'manual':
            # For manual merges, combine both states (a takes precedence for conflicts)
            state_b = dict(block_b.get('state', {})) if block_b else {}
            state_a = dict(block_a.get('state', {})) if block_a else {}
            state = {**state_b, **state_a}
        else:
            raise ValueError(f"FoodBlock: unknown merge strategy '{strategy}'")

    conflict = detect_conflict(hash_a, hash_b, resolve)

    merge_state = {
        **state,
        '_merge': {
            'strategy': strategy,
            'is_conflict': conflict['is_conflict'],
            'common_ancestor': conflict['common_ancestor'],
        },
    }

    return create('observe.merge', merge_state, {
        'merge_a': hash_a,
        'merge_b': hash_b,
    })


def auto_merge(hash_a, hash_b, resolve, vocabulary=None):
    """
    Automatically merge two forked chains using field-level resolution.

    If a vocabulary is provided, required fields from chain A take precedence,
    and optional fields are merged. Without a vocabulary, falls back to
    combining both states with A winning on conflicts.

    Args:
        hash_a: hash of chain A head
        hash_b: hash of chain B head
        resolve: callable (hash) -> block dict or None
        vocabulary: optional vocabulary block dict for field-aware merging

    Returns:
        An observe.merge FoodBlock dict
    """
    block_a = resolve(hash_a)
    block_b = resolve(hash_b)

    state_a = dict(block_a.get('state', {})) if block_a else {}
    state_b = dict(block_b.get('state', {})) if block_b else {}

    if vocabulary:
        fields = vocabulary.get('state', vocabulary).get('fields', {})
        merged_state = {}

        # Start with all of B's state
        merged_state.update(state_b)

        # Override with A's state
        merged_state.update(state_a)

        # For fields defined in vocabulary, apply smarter merging
        for field_name, field_def in fields.items():
            a_val = state_a.get(field_name)
            b_val = state_b.get(field_name)

            if a_val is not None and b_val is not None:
                # Both have the field — required fields: A wins; optional: prefer non-empty
                if field_def.get('required'):
                    merged_state[field_name] = a_val
                else:
                    # Prefer the more "complete" value
                    merged_state[field_name] = a_val if a_val else b_val
            elif a_val is not None:
                merged_state[field_name] = a_val
            elif b_val is not None:
                merged_state[field_name] = b_val
    else:
        # No vocabulary — simple merge, A wins on conflicts
        merged_state = {**state_b, **state_a}

    return merge(hash_a, hash_b, resolve, state=merged_state, strategy='manual')


def _walk_chain(start_hash, resolve, max_depth=100):
    """Walk an update chain backwards, returning list of blocks."""
    visited = set()
    result = []
    current = start_hash
    depth = 0

    while current and depth < max_depth:
        if current in visited:
            break
        visited.add(current)

        block = resolve(current)
        if not block:
            break

        result.append(block)

        refs = block.get('refs', {})
        updates = refs.get('updates')
        if isinstance(updates, list):
            current = updates[0] if updates else None
        else:
            current = updates

        depth += 1

    return result


def _trim_to_ancestor(chain_blocks, ancestor_hash):
    """Trim a chain to include blocks up to and including the ancestor."""
    if ancestor_hash is None:
        return list(chain_blocks)

    trimmed = []
    for block in chain_blocks:
        trimmed.append(block)
        if block['hash'] == ancestor_hash:
            break
    return trimmed

File: sdk/swift/Sources/Merge.swift
import Foundation

// MARK: - Types

/// Result of conflict detection between two chain heads.
public struct ConflictResult {
    public let isConflict: Bool
    public let commonAncestor: String?
    public let chainA: [FoodBlock]
    public let chainB: [FoodBlock]
}

/// Errors that can occur during merge operations.
public enum MergeError: Error, CustomStringConvertible {
    case manualMergeRequiresState
    case couldNotResolve(String)
    case unknownStrategy(String)
    case autoMergeConflict(String)
    case unknownFieldStrategy(String, String)

    public var description: String {
        switch self {
        case .manualMergeRequiresState:
            return "FoodBlock: manual merge requires state"
        case .couldNotResolve(let hash):
            return "FoodBlock: could not resolve \(hash)"
        case .unknownStrategy(let strategy):
            return "FoodBlock: unknown merge strategy: \(strategy)"
        case .autoMergeConflict(let field):
            return "FoodBlock: auto-merge conflict on field \"\(field)\" — manual resolution required"
        case .unknownFieldStrategy(let strategy, let field):
            return "FoodBlock: unknown merge strategy \"\(strategy)\" for field \"\(field)\""
        }
    }
}

// MARK: - FoodBlockMerge

/// Fork detection and conflict resolution for forked update chains.
/// Merges turn chains into DAGs.
public enum FoodBlockMerge {

    /// Extract the `updates` ref from a block, handling both String and Array forms.
    private static func updatesRef(from block: FoodBlock) -> String? {
        guard let updates = block.refs["updates"] else { return nil }
        if let str = updates.value as? String {
            return str
        }
        if let arr = updates.value as? [Any], let first = arr.first as? String {
            return first
        }
        return nil
    }

    /// Detect whether two block hashes represent a fork (conflict) in an
    /// update chain by walking both chains backward to find a common ancestor.
    ///
    /// - Parameters:
    ///   - hashA: First chain head hash.
    ///   - hashB: Second chain head hash.
    ///   - resolve: Async function that resolves a hash to a FoodBlock.
    /// - Returns: A `ConflictResult` indicating whether a conflict exists.
    public static func detectConflict(
        hashA: String,
        hashB: String,
        resolve: @escaping (String) async -> FoodBlock?
    ) async -> ConflictResult {
        if hashA == hashB {
            return ConflictResult(isConflict: false, commonAncestor: hashA, chainA: [], chainB: [])
        }

        var chainA: [FoodBlock] = []
        var visitedA = Set<String>()

        // Walk chain A backwards
        var current: String? = hashA
        while let hash = current {
            visitedA.insert(hash)
            guard let block = await resolve(hash) else { break }
            chainA.append(block)
            current = updatesRef(from: block)
        }

        // Walk chain B backwards, looking for intersection with A
        var chainB: [FoodBlock] = []
        var commonAncestor: String? = nil
        current = hashB
        while let hash = current {
            if visitedA.contains(hash) {
                commonAncestor = hash
                break
            }
            guard let block = await resolve(hash) else { break }
            chainB.append(block)
            current = updatesRef(from: block)
        }

        return ConflictResult(
            isConflict: commonAncestor != nil,
            commonAncestor: commonAncestor,
            chainA: chainA,
            chainB: chainB
        )
    }

    /// Create a merge block that resolves a fork between two chain heads.
    ///
    /// - Parameters:
    ///   - hashA: First fork head hash.
    ///   - hashB: Second fork head hash.
    ///   - resolve: Async function that resolves a hash to a FoodBlock.
    ///   - strategy: Merge strategy: "manual", "a_wins", or "b_wins".
    ///   - state: Required when strategy is "manual" — the manually resolved state.
    /// - Returns: An observe.merge FoodBlock.
    /// - Throws: `MergeError` if the strategy is invalid or required data is missing.
    public static func merge(
        hashA: String,
        hashB: String,
        resolve: @escaping (String) async -> FoodBlock?,
        strategy: String = "manual",
        state: [String: Any]? = nil
    ) async throws -> FoodBlock {
        let mergedState: [String: Any]

        switch strategy {
        case "manual":
            guard let manualState = state else {
                throw MergeError.manualMergeRequiresState
            }
            mergedState = manualState

        case "a_wins":
            guard let blockA = await resolve(hashA) else {
                throw MergeError.couldNotResolve("hashA")
            }
            mergedState = blockA.state.mapValues { $0.value }

        case "b_wins":
            guard let blockB = await resolve(hashB) else {
                throw MergeError.couldNotResolve("hashB")
            }
            mergedState = blockB.state.mapValues { $0.value }

        default:
            throw MergeError.unknownStrategy(strategy)
        }

        var finalState: [String: Any] = ["strategy": strategy]
        for (key, value) in mergedState {
            finalState[key] = value
        }

        return FoodBlock.create(
            type: "observe.merge",
            state: finalState,
            refs: ["merges": [hashA, hashB]]
        )
    }

    /// Attempt an automatic merge using vocabulary-defined per-field strategies.
    ///
    /// Supported per-field strategies (from vocabulary fields[key].merge):
    /// - "lww" / "last_writer_wins": prefer value from B (convention: later writer)
    /// - "max": take the greater numeric value
    /// - "min": take the lesser numeric value
    /// - "union": merge arrays, deduplicating values
    /// - "conflict": always raise an error (default when no strategy specified)
    ///
    /// - Parameters:
    ///   - hashA: First fork head hash.
    ///   - hashB: Second fork head hash.
    ///   - resolve: Async function that resolves a hash to a FoodBlock.
    ///   - vocabulary: Optional vocabulary with per-field merge strategies.
    ///                 Expected shape: `["fields": ["fieldName": ["merge": "lww"]]]`
    /// - Returns: An observe.merge FoodBlock with auto-merged state.
    /// - Throws: `MergeError` if a field conflict cannot be resolved.
    public static func autoMerge(
        hashA: String,
        hashB: String,
        resolve: @escaping (String) async -> FoodBlock?,
        vocabulary: [String: Any]? = nil
    ) async throws -> FoodBlock {
        guard let blockA = await resolve(hashA) else {
            throw MergeError.couldNotResolve("hashA")
        }
        guard let blockB = await resolve(hashB) else {
            throw MergeError.couldNotResolve("hashB")
        }

        let stateA = blockA.state.mapValues { $0.value }
        let stateB = blockB.state.mapValues { $0.value }

        // Gather all keys from both states
        let allKeys = Set(Array(stateA.keys) + Array(stateB.keys))
        var mergedState: [String: Any] = [:]

        // Extract fields definition from vocabulary
        let fields: [String: Any]?
        if let vocab = vocabulary {
            if let vocabState = vocab["state"] as? [String: Any],
               let f = vocabState["fields"] as? [String: Any] {
                fields = f
            } else if let f = vocab["fields"] as? [String: Any] {
                fields = f
            } else {
                fields = nil
            }
        } else {
            fields = nil
        }

        for key in allKeys {
            let valA = stateA[key]
            let valB = stateB[key]

            // If values are the same (by canonical JSON comparison), no conflict
            if canonicalEqual(valA, valB) {
                mergedState[key] = valA ?? valB
                continue
            }

            // If only one side has the value, take it
            if valA == nil { mergedState[key] = valB; continue }
            if valB == nil { mergedState[key] = valA; continue }

            // Values differ -- use vocabulary strategy if available
            let fieldDef = (fields?[key] as? [String: Any])
            let mergeStrategy = fieldDef?["merge"] as? String

            guard let strategy = mergeStrategy, strategy != "conflict" else {
                throw MergeError.autoMergeConflict(key)
            }

            switch strategy {
            case "last_writer_wins", "lww":
                // Prefer B (convention: later writer)
                mergedState[key] = valB

            case "max":
                if let numA = asDouble(valA), let numB = asDouble(valB) {
                    mergedState[key] = Swift.max(numA, numB)
                } else {
                    mergedState[key] = valB
                }

            case "min":
                if let numA = asDouble(valA), let numB = asDouble(valB) {
                    mergedState[key] = Swift.min(numA, numB)
                } else {
                    mergedState[key] = valB
                }

            case "union":
                let arrA = asArray(valA)
                let arrB = asArray(valB)
                mergedState[key] = unionArrays(arrA, arrB)

            default:
                throw MergeError.unknownFieldStrategy(strategy, key)
            }
        }

        var finalState: [String: Any] = ["strategy": "auto"]
        for (key, value) in mergedState {
            finalState[key] = value
        }

        return FoodBlock.create(
            type: "observe.merge",
            state: finalState,
            refs: ["merges": [hashA, hashB]]
        )
    }

    // MARK: - Private Helpers

    /// Compare two values for canonical equality using their string representations.
    private static func canonicalEqual(_ a: Any?, _ b: Any?) -> Bool {
        if a == nil && b == nil { return true }
        guard let a = a, let b = b else { return false }
        return canonicalString(a) == canonicalString(b)
    }

    /// Produce a canonical string representation of a value for comparison.
    private static func canonicalString(_ value: Any) -> String {
        if let str = value as? String { return str }
        if let num = value as? Int { return "\(num)" }
        if let num = value as? Double { return "\(num)" }
        if let bool = value as? Bool { return bool ? "true" : "false" }
        if let arr = value as? [Any] {
            let parts = arr.map { canonicalString($0) }
            return "[\(parts.joined(separator: ","))]"
        }
        if let dict = value as? [String: Any] {
            let sorted = dict.keys.sorted().map { key in
                "\(key):\(canonicalString(dict[key]!))"
            }
            return "{\(sorted.joined(separator: ","))}"
        }
        return String(describing: value)
    }

    /// Attempt to extract a Double from any value.
    private static func asDouble(_ value: Any?) -> Double? {
        guard let value = value else { return nil }
        if let d = value as? Double { return d }
        if let i = value as? Int { return Double(i) }
        if let n = value as? NSNumber { return n.doubleValue }
        return nil
    }

    /// Coerce a value into an array. If already an array, return it; otherwise wrap in one.
    private static func asArray(_ value: Any?) -> [Any] {
        guard let value = value else { return [] }
        if let arr = value as? [Any] { return arr }
        return [value]
    }

    /// Union two arrays, deduplicating by canonical string representation.
    private static func unionArrays(_ a: [Any], _ b: [Any]) -> [Any] {
        var seen = Set<String>()
        var result: [Any] = []
        for item in a + b {
            let key = canonicalString(item)
            if !seen.contains(key) {
                seen.insert(key)
                result.append(item)
            }
        }
        return result
    }
}
